Give TiledBacking a TiledBackingClient that allows a client more knowledge of tile painting

smfr opened the pull request:
#### 3f6d7f5ff24f92cb8327e9d67a87755d6d3d47a4
<pre>
Give TiledBacking a TiledBackingClient that allows a client more knowledge of tile painting
<a href="https://bugs.webkit.org/show_bug.cgi?id=269268">https://bugs.webkit.org/show_bug.cgi?id=269268</a>
<a href="https://rdar.apple.com/122850594">rdar://122850594</a>

Reviewed by Tim Horton.

Introduce TiledBackingClient, a weakly held client of TiledBacking that gets informed
when a tile is repainted or removed, and when all tiles are invalidate.

UnifiedPDFPlugin will use this as a basis for caching.

The client is informed about events per tile using TileGridIndex and TileIndex to
identify tiles; TileGridIndex is 0 for the main tile grid, and 1 for the optional zoomed-out
tile grid.

* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setClient):
(WebCore::TileController::willRepaintTile):
(WebCore::TileController::willRemoveTile):
(WebCore::TileController::willRepaintAllTiles):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::setScale):
(WebCore::TileGrid::setNeedsDisplay):
(WebCore::TileGrid::setTileNeedsDisplayInRect):
(WebCore::TileGrid::removeTiles):
(WebCore::TileGrid::revalidateTiles):
(WebCore::TileGrid::ensureTilesForRect):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:

Canonical link: <a href="https://commits.webkit.org/274574@main">https://commits.webkit.org/274574@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcf0357f8dada045db1825569dbf19decc37934d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18429 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41804 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41984 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35350 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41756 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21306 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15758 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32978 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40024 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15524 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34169 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13487 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13457 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35115 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43262 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35818 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35445 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39254 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14262 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37505 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15868 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8834 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15916 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/ca/TileController.h:
```diff
@@ -46,6 +46,8 @@ class IntRect;
 class TileCoverageMap;
 class TileGrid;
 
+using TileIndex = IntPoint;
```
This is defined in `TiledBacking.h`. Why do you need this?

smfr committed:
Give TiledBacking a TiledBackingClient that allows a client more knowledge of tile painting
https://bugs.webkit.org/show_bug.cgi?id=269268
rdar://122850594

Reviewed by Tim Horton.

Introduce TiledBackingClient, a weakly held client of TiledBacking that gets informed
when a tile is repainted or removed, and when all tiles are invalidate.

UnifiedPDFPlugin will use this as a basis for caching.

The client is informed about events per tile using TileGridIndex and TileIndex to
identify tiles; TileGridIndex is 0 for the main tile grid, and 1 for the optional zoomed-out
tile grid.

* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setClient):
(WebCore::TileController::willRepaintTile):
(WebCore::TileController::willRemoveTile):
(WebCore::TileController::willRepaintAllTiles):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::setScale):
(WebCore::TileGrid::setNeedsDisplay):
(WebCore::TileGrid::setTileNeedsDisplayInRect):
(WebCore::TileGrid::removeTiles):
(WebCore::TileGrid::revalidateTiles):
(WebCore::TileGrid::ensureTilesForRect):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:

Canonical link: https://commits.webkit.org/274574@main
